### ISSUE TITLE
Assertion failure while waking from sleep on MacOS

### ISSUE THREAD

**Issue by Guistac** [ocornut/imgui#5683]
Version: 1.89 (Docking)
Back-ends: `imgui_impl_glfw` `imgui_impl_opengl3`
OS: MacOS 12.3.1

When waking my machine from sleep while running my app, imgui fails to assert `g.PlatformIO.Monitors.Size > 0`
https://github.com/ocornut/imgui/blob/64b88da21ba554eb5735a387d91d665cd387afdf/imgui.cpp#L8938
It doesn't do this consistently, there is about a 50% chance it will happen on each sleep wake cycle.
How does disabling this assertion affect the integrity of imgui?
I've currently commented it out and experienced no issues.
I'm curious if anyone experiences the same issue, as I'm currently unable to test it on another system.

**Comment by ocornut** [comment]
This is needed for multi-viewport support.
Instead of disabling the assert we should be investigating what leads the GLFW backends to clear this and not restore it in time. Adding logging (withf
frame count) in the GLFW events/functions leading to change of monitor data would be useful.

**Comment by ocornut** [comment]
This is needed for multi-viewport support.
Instead of disabling the assert we should be investigating what leads the GLFW backends to clear this and not restore it in time. Adding logging (withf
frame count) in the GLFW events/functions leading to change of monitor data would be useful.

**Comment by PathogenDavid** [comment]
This might just be a quirk of [how macOS handles monitors during standby](https://notes.alinpanaitiu.com/Weird%20monitor%20bugs#how-monitor-settings-are-stored):

> In standby, the monitor is actually disconnected and disappears from the I/O Registry, so on system wake, it reconnects and the OS has to rearrange based on the stored settings.

GLFW on macOS [filters out displays which are asleep](https://github.com/glfw/glfw/blob/955fbd9d265fa95adf9cb94896eb9a516aa50420/src/cocoa_monitor.m#L318-L319). This seems reasonable, but according to the quirk described above that sounds like that creates a brief moment in time where GLFW could report 0 monitors.

Ignoring the assert is definitely wrong, but maybe it'd be reasonable for the GLFW backend to just ignore the monitor change event when `glfwGetMonitors` returns no monitors?

**Comment by ocornut** [comment]
> Ignoring the assert is definitely wrong, but maybe it'd be reasonable for the GLFW backend to just ignore the monitor change event when glfwGetMonitors returns no monitors?

I agree, that sounds good (with a comment and # ref to explain why)

**Comment by phraktle** [comment]
Any plans for a fix? This does happen still.

**Comment by emoon** [comment]
I don't want to hijack this issue, but I'm getting the same assert if you try to use the `glfw_opengl3` example with Emscripten on the docking branch (regular works fine)
```
Uncaught RuntimeError: Aborted(Assertion failed: g.PlatformIO.Monitors.Size > 0 && "Platform init didn't setup Monitors list?", at: ../../imgui.cpp,9623,ErrorCheckNewFrameSanityChecks)
```

**Comment by Guistac** [comment]
Hi, I did manage to fix this particular issue by adding the following code to `ImGui_ImplGlfw_UpdateMonitors()` right after
```GLFWmonitor** glfw_monitors = glfwGetMonitors(&monitors_count);```

```
#if __APPLE__
//waking MacOS from sleep sometimes causes glfw to report 0 monitors.
//if this happens we abort updating monitors but keep WantUpdateMonitors == true to check again on the next frame
//git issue: https://github.com/ocornut/imgui/issues/5683
if(monitors_count == 0){
  bd->WantUpdateMonitors = true;
  return;
}
#endif
```

I wanted to do a pull request but I'm pretty new to developing and got quite confused by how git works...

**Comment by phraktle** [comment]
Happens on my MacBook Pro (15-inch, 2016), MacOS Monterey 12.6.3 when waking from sleep (though not deterministically). I did not yet see it on a more recent M1 mac.

**Comment by phraktle** [comment]
happens on the M2 as well, but less frequently.

**Comment by JaedanC** [comment]
Sorry to bring up an old issue but this commit [`19b436f`](https://github.com/ocornut/imgui/commit/19b436f68da9ceaef3684900c90ec7a9622f941a) actually breaks my code and results in the `IM_ASSERT(g.PlatformIO.Monitors.Size > 0 && "Platform init didn't setup Monitors list?");` being triggered on the first call to `NewFrame`. I'm still investigating what's going on and I'll make a separate issue once I figure it out.

**Comment by ocornut** [comment]
Apologies for the typo, pushed fix 2e810d575cb5c8708340f53e301424c237ebca80 (first time I do a code edit directly on github!).
